### PR TITLE
schema: Unify schema hierarchy

### PR DESCRIPTION
**Pull request by tyll**
Move constant to LinuxBridge.Port, LinuxBridge.STP and OVSBridge.Port
classes. Create DeprecationWarnings when the old constants are used.

Signed-off-by: Till Maas <opensource@till.name>

@EdDev this is an example about how we could change/deprecate the constants. I did this only for the LinuxBridge.Port constants for now but if you agree with this approach, it can easily be done for the other candidates as well. Also change the constants in our CI is missing and the API guide needs to be updated afterwards.

---
This change is [Reviewable](https://reviewable.io/reviews/nmstate/nmstate/608)

### CODE REVIEW

**Comment by EdDev** [comment]
After playing with this a little bit, here is my attempt to generalize it.

Regardless, I suggest that the new schema constants will get introduced through a separate PR (or commit). The warning is less important than the actual new changes it will allow us to adjust each part independently.

The below code may be upgraded further by allowing each module to register deprecation attributes per class. This may be handy for the long run as well as an infrastructure to declare such things (but we do not have to leave it, just an option).

```python
_depricate_map = {'Foo': {'FOO': 'Boo.FOO'}}


class DepricatorType(type):
    def __getattribute__(cls, attr):
        val = super().__getattribute__(attr)
        new_attr = _depricate_map.get(super().__name__, {}).get(attr)
        if new_attr:
            warnings.warn(
                f'{attr} deprecated by {new_attr}',
                FutureWarning,
                stacklevel=3,
            )
            new_val = super().__getattribute__(new_attr.split('.', 1)[0])
            return new_val
        return val


class Foo(metaclass=DepricatorType):
    FOO = 'foo'

    class Boo:
        FOO = 'foo'
```

**Comment by tyll** [comment]
> After playing with this a little bit, here is my attempt to generalize it.
> 
> Regardless, I suggest that the new schema constants will get introduced through a separate PR (or commit). The warning is less important than the actual new changes it will allow us to adjust each part independently.

This means to do meaningless work on creating tests of code that needs to be replaced by Sunday anyhow to add the warnings. The warnings need to be there to properly document the situation in the next release.
 
> The below code may be upgraded further by allowing each module to register deprecation attributes per class. This may be handy for the long run as well as an infrastructure to declare such things (but we do not have to leave it, just an option).

Can you please describe what you think is better with this approach except for handling a situation that we do not need now but might need in the future? I added my comments as python comments in your code snippet:

> ```python
> _depricate_map = {'Foo': {'FOO': 'Boo.FOO'}}  # IMHO this is more error-prone because the names need to be specified multiple times and if there is a typo, it would not be catched. Also it needs to keep three places in sync (this dictionary, the old constant and the new constant) In my proposal, the new attribute and the old ones are mentioned exactly once. 
> 
> 
> class DepricatorType(type):
>     def __getattribute__(cls, attr):
>         val = super().__getattribute__(attr)
>         new_attr = _depricate_map.get(super().__name__, {}).get(attr)
>         if new_attr:
>             warnings.warn(
>                 f'{attr} deprecated by {new_attr}',  # The warning is missing the class names, it says FOO deprecated by Boo.FOO instead of Foo.FOO and Foo.Boo.FOO
>                 FutureWarning,
>                 stacklevel=3,
>             )
>             new_val = super().__getattribute__(new_attr.split('.', 1)[0])
>             return new_val
>         return val
> 
> 
> class Foo(metaclass=DepricatorType):
>     FOO = 'foo'  # The advantage of my proposal is that pylint warns about the deprecated members because the do not appear here. This makes it easier to catch code that still uses the deprecated values. This proposal does not do this, see https://travis-ci.com/nmstate/nmstate/jobs/261048861#L850 - also my proposal ensures that both constants have the same value because the value exists only once.
> 
>     class Boo:
>         FOO = 'foo'
> ```

**Comment by EdDev** [comment]
> This means to do meaningless work on creating tests of code that needs to be replaced by Sunday anyhow to add the warnings. The warnings need to be there to properly document the situation in the next release.

For the added consts which are not yet used you do not need tests. The existing tests protect that we have not done anything bad. When we move to use the new constants, the tests should protect us.

> Can you please describe what you think is better with this approach except for handling a situation that we do not need now but might need in the future? I added my comments as python comments in your code snippet

Simpler, less code, generic and not specific per constant class, can be re-used on any constant.
I thought it is obvious when looking at it, unless I totally missed here something critical.

**Comment by EdDev** [comment]
> The advantage of my proposal is that pylint warns about the deprecated members because the do not appear here. This makes it easier to catch code that still uses the deprecated values. This proposal does not do this, see https://travis-ci.com/nmstate/nmstate/jobs/261048861#L850 - also my proposal ensures that both constants have the same value because the value exists only once.

- warning logs are raised, seems enough for me.
- It is may be possible to to add an assert to check that the values are equal, but that can be done on definition or not at all by trusting the tests we already have.

**Comment by tyll** [comment]
> > This means to do meaningless work on creating tests of code that needs to be replaced by Sunday anyhow to add the warnings. The warnings need to be there to properly document the situation in the next release.
> 
> For the added consts which are not yet used you do not need tests. The existing tests protect that we have not done anything bad. When we move to use the new constants, the tests should protect us.

Why are the added constants special, that they do not need tests? They are part of the public API so it is IMHO rather important that they are correct. 

> > Can you please describe what you think is better with this approach except for handling a situation that we do not need now but might need in the future? I added my comments as python comments in your code snippet
> 
> Simpler, less code, generic and not specific per constant class, can be re-used on any constant.
> I thought it is obvious when looking at it, unless I totally missed here something critical.

Not sure what you mean with simpler. To me it is harder to use and understand because it hides the information that attributes are deprecated in an unrelated dictionary and instead of showing it in the `Foo` class that does not allow to see which constant is the correct one. In my example it uses a easy to understand decorators to see that for example `PORT_NAME` is a property and deprecated. It might be less code but therefore more duplication by having the same identifiers at multiple places.
IMHO in your approach, instead of using Python code to express the intentions, using the dictionary seems to me like inventing a custom syntax and programming with the dictionary to save some lines of code but by increasing the redundancy (in a bad way). 

> > The advantage of my proposal is that pylint warns about the deprecated members because the do not appear here. This makes it easier to catch code that still uses the deprecated values. This proposal does not do this, see https://travis-ci.com/nmstate/nmstate/jobs/261048861#L850 - also my proposal ensures that both constants have the same value because the value exists only once.
> 
> * warning logs are raised, seems enough for me.

The logs might only be shown when the slow tests that cover them run. Pylint is fast and allows to spot the mistakes while developing with integration in editors.

> * It is may be possible to to add an assert to check that the values are equal, but that can be done on definition or not at all by trusting the tests we already have.

In my proposal, each value exists only once, it is IMHO an easier approach to ensure that the values are the same instead of writing them down twice and adding extra asserts.

**Comment by tyll** [comment]
By the way, it is still possible to avoid the redundancy and have a generic type. This example also allows to move constants between classes. Still not sure that using dictionaries for this is a nicer way. I am only sure that this is overengineering for Nmstate and seems to becoming code for a deprecation module project. I do not really see the extra value for Nmstate compared to my other approach since both pass the tests.


```python
DEPRECATED_CONSTANTS = {
    'LinuxBridge.PORT_NAME': 'LinuxBridge.Port.NAME',
    'LinuxBridge.PORT_STP_PRIORITY': 'LinuxBridge.Port.STP_PRIORITY',
    'LinuxBridge.PORT_STP_HAIRPIN_MODE': 'LinuxBridge.Port.STP_HAIRPIN_MODE',
    'LinuxBridge.PORT_STP_PATH_COST': 'LinuxBridge.Port.STP_PATH_COST',
}


class _NmstateContants(type):
    def __getattribute__(cls, attribute):
        try:
            return super().__getattribute__(attribute)
        except AttributeError:
            deprecated_class = cls
            deprecated_classname = deprecated_class.__name__
            deprecated_name = attribute

            oldconstant = f'{deprecated_classname}.{deprecated_name}'
            newconstant = DEPRECATED_CONSTANTS.get(oldconstant)

            if newconstant:
                _warn_deprecation(
                    f"Using '{oldconstant}' is deprecated, "
                    f"use '{newconstant}' instead."
                )

                attributes = newconstant.split('.')
                new_classname = attributes.pop(0)
                new_value = globals()[new_classname]
                while attributes:
                    new_value = getattr(new_value, attributes.pop(0))

                return new_value

            raise


class LinuxBridge(metaclass=_NmstateContants):
    class Port(object):
        NAME = 'name'
        STP_HAIRPIN_MODE = 'stp-hairpin-mode'
        STP_PATH_COST = 'stp-path-cost'
        STP_PRIORITY = 'stp-priority'
```

**Comment by EdDev** [comment]
> Why are the added constants special, that they do not need tests? They are part of the public API so it is IMHO rather important that they are correct.

Because they have not been used yet, not in the production code, not in the tests and not by anyone else.

> Not sure what you mean with simpler.

You have used a metaclass to mark a specific class in a 1:1 relationship. That is IMO not simple, hard to follow and too specific for something which has a generic need.
We probably need to create such indented classes all over, even for the existing code, and with your approach it will require a lot of code. This was the point that made me uncomfortable when I first saw it.

>  It might be less code but therefore more duplication by having the same identifiers at multiple places.

I did not understood this one.

> IMHO in your approach, instead of using Python code to express the intentions, using the dictionary seems to me like inventing a custom syntax and programming with the dictionary to save some lines of code but by increasing the redundancy (in a bad way).

You can use an object instead of a dict, it does not matter. It is just a data structure.
Data structure is IMO better then havind dedicated metaclasses with dedicated methods that track the original class.
I also do not know what "redundancy" you refer to.

> In my proposal, each value exists only once, it is IMHO an easier approach to ensure that the values are the same instead of writing them down twice and adding extra asserts.

That's a technicality. It's fine to use in the new definition the old definition value:
`NEWFOO = OLDFOO`
That will do the same, but again this is why the existing tests are in place.

**Comment by EdDev** [comment]
Just now saw your new proposal.
That one looks nice from the first look, thanks.

I still think the commits should be separated, first adding the new constants, then adding the warning.
After that, we can start change the internal code of nmstate, and lastly the the tests.

**Comment by tyll** [comment]
> > Why are the added constants special, that they do not need tests? They are part of the public API so it is IMHO rather important that they are correct.
> 
> Because they have not been used yet, not in the production code, not in the tests and not by anyone else.

I do not follow this reasoning. All new code will not be used in any production code, yet. And of course, when we do not add tests, it will not be used in tests...

> > Not sure what you mean with simpler.
> 
> You have used a metaclass to mark a specific class in a 1:1 relationship. That is IMO not simple, hard to follow and too specific for something which has a generic need.
> We probably need to create such indented classes all over, even for the existing code, and with your approach it will require a lot of code. This was the point that made me uncomfortable when I first saw it.

Why is there a generic need? We do not have a generic need, we have a specific need now.

> > It might be less code but therefore more duplication by having the same identifiers at multiple places.
> 
> I did not understood this one.
> 
> > IMHO in your approach, instead of using Python code to express the intentions, using the dictionary seems to me like inventing a custom syntax and programming with the dictionary to save some lines of code but by increasing the redundancy (in a bad way).
> 
> You can use an object instead of a dict, it does not matter. It is just a data structure.
> Data structure is IMO better then havind dedicated metaclasses with dedicated methods that track the original class.

The python syntax to add a property is to have a method and decorate it with `@property`. This is what happened in my original code. By using `__getattr__` with a custom datastructure, it is happening in a custom way (currently a dictionary) instead of using the the Python way that is easier to understand.

> I also do not know what "redundancy" you refer to.

In your example, the constants exist multiple times, in the Foo class and also in the dictionary. In my original example, there was only the new constant and one property marked as deprecated without copying the same information to multiple places.

> > In my proposal, each value exists only once, it is IMHO an easier approach to ensure that the values are the same instead of writing them down twice and adding extra asserts.
> 
> That's a technicality. It's fine to use in the new definition the old definition value:
> `NEWFOO = OLDFOO`

ok, then propose it like this. Also after another look, it seems that the `_getattr__` implementation actually takes the new value instead of the original value which is a result of the redundancy making the code less obvious.

> That will do the same, but again this is why the existing tests are in place.

AFAIK, if the constants are used, its the same in the lib and in the tests, so the tests would just test that the new constants equal the new constants. To test it correctly. the tests would need to use their own copy of constants that they would expect regardless of the internal implementation of whether the constants are called `PORT_NAME` or `Port.NAME`.

**Comment by tyll** [comment]
> I still think the commits should be separated, first adding the new constants, then adding the warning.
> After that, we can start change the internal code of nmstate, and lastly the the tests.
> But I will not fight too much over it.

Why do you prefer interim steps that leave the repository in a broken/inconsistent state? If the new constants are added it is IMHO correct to deprecate the old values and adjust the tests to keep the code consistent.

**Comment by tyll** [comment]
Here is now a full example for the LinuxBridge Port constants. Missing is at least LinuxBridge STP and OvsBridge Port. I noticed there are also other both also have an options subtree (not sure why it is plural but Port is singular) that could be moved to classes as well.

Since this PR contains also the interim commits, hope you will approve this PR. Then I can also add follow-up PRs for LinuxBridge.STP and OvsBridge.Port - what is your opinion on the options subtrees?

File: libnmstate/schema.py [hunk 1]
```diff
@@ -158,7 +162,46 @@ class BondMode(object):
...
-class LinuxBridge(object):
+DEPRECATED_CONSTANTS = {
+    'LinuxBridge.PORT_NAME': 'LinuxBridge.Port.NAME',
+    'LinuxBridge.PORT_STP_PRIORITY': 'LinuxBridge.Port.STP_PRIORITY',
+    'LinuxBridge.PORT_STP_HAIRPIN_MODE': 'LinuxBridge.Port.STP_HAIRPIN_MODE',
+    'LinuxBridge.PORT_STP_PATH_COST': 'LinuxBridge.Port.STP_PATH_COST',
+}
+
+
+class _NmstateContants(type):
```
[the remark this answers is not in the document]
Comment (tyll) [reply libnmstate/schema.py]: The class/type itself does not contain any information about deprecation, it is all in the separate dictionary. It can be used for all constants, even if they do not have anything that is deprecated. If there is another common code for all constant classes, it could go in there as well.
However, I do not care too much about the name, just wanted to share my thoughts about this.

File: libnmstate/schema.py [hunk 2]
```diff
@@ -158,7 +162,46 @@ class BondMode(object):
...
+            deprecated_name = attribute
+
+            oldconstant = f'{deprecated_classname}.{deprecated_name}'
+            newconstant = DEPRECATED_CONSTANTS.get(oldconstant)
+
+            if newconstant:
+                warnings.warn(
+                    f"Using '{oldconstant}' is deprecated, "
+                    f"use '{newconstant}' instead.",
+                    DeprecationWarning,
```
[the remark this answers is not in the document]
Comment (tyll) [reply libnmstate/schema.py]: Works for me, will change it. We can also add a warning via logging if you like.

**Comment by tyll** [comment]
> Commit: `Schema: Deprecate old LinuxBridge PORT_ constants`
> Please elaborate and explain the deprecation handling.

Done, including the other requests.

**Comment by tyll** [comment]
@EdDev please share your opinion regarding the options subtrees.

**Comment by tyll** [comment]
Also avoided to use `LB` in the new test and sorted the deprecation dictionary.

**Comment by EdDev** [comment]
What about the others now? Any chance it will be done before the new version?

**Comment by tyll** [comment]
> What about the others now? Any chance it will be done before the new version?

#612 contains the two others we already discussed (LinuxBridge.STP and OVSBridge.Port). I still need your feedback regarding the options subtrees.

**Comment by EdDev** [comment]
If you mean https://github.com/nmstate/nmstate/blob/b8e180851a9c6f7ed778c93b34656dc80810531f/libnmstate/schema.py#L208 then I would say yes, it should be indented.

I would argue that all should follow the schema hierarchy for consistency.
To make it reasonable consumable, we could do this: Convert the current schema module to a sub-package and create as modules the top level subtrees (interface, route, dns...).
